qemudriver: Allow machine-specific options and pre-start QMP commands

JSydll opened the pull request:
**Description**

Two minor features are added to the `QemuDriver`:

_Support machine-specific options_ (e.g. `-machine virt,secure=on`)

> While raising an exception for unsupported machines/boards is reasonable, adding machine-specific options shall not break the parsing. Therefore, the check is adjusted to only account for the leading board name.

_Support interactions with the QEMU instance via QMP before releasing CPU(s)_

> Especially for bad case testing, it can be useful to be able to interact with the prepared but not yet spun off instance - e.g. to corrupt the boot medium without the need to build a corrupted image only for that purpose.

For both features, straight forward unit tests are provided.
Lastly, the changeset includes minor documentation updates and the introduction of type hints for the `qemudriver.py` file.

**Checklist**
- [x] Documentation for the feature
- [x] Tests for the feature 
- [x] PR has been tested

JSydll committed:
qemudriver: Allow machine-specific options.

Signed-off-by: Joschka Seydell <joschka@seydell.org>

JSydll committed:
qemudriver: Support interactions via monitor_command before spinning off the instance.

Signed-off-by: Joschka Seydell <joschka@seydell.org>

JSydll committed:
dockerfiles: Add missing VERSION parameter for building containers.

Signed-off-by: Joschka Seydell <joschka@seydell.org>

JSydll committed:
qemudriver: Add type hints.

Signed-off-by: Joschka Seydell <joschka@seydell.org>

Emantor reviewed (commented):
Rather than adding a pre-start hook, I think it is more clear to expose somthing like a `prepare()` function which sets up the qemu process, and than have `on()` only do the `cont` monitor command. For backwards compatibility, the class can check whether the `self._child` attribute exists and call `prepare()` in `on()` if that was not done yet.

Emantor reviewed (dismissed):
I think we can take the other fixes in this PR as well.

Emantor commented:
It's all good and thanks for keeping up with the PR even if it takes some time on our side :+1:

JSydll commented:
I shortly glimpsed into the failing pipeline and now I'm wondering whether some environment changes happened? The test_qemudriver is not able to locate qemu-system-arm...

Emantor committed:
github: add qemu to reusable unit

In order to build the new tests for the qemu driver, the qemu-system-arm
binary needs to be available on the system.

Signed-off-by: Rouven Czerwinski <rouven@czerwinskis.de>

Bastian-Krause commented:
Prior to this, we've mocked qemu. Why do we need to use the real binary now?

If we decide we need the real qemu, we should at least add a marker so the tests get skipped on systems without it. Something along the lines of:

https://github.com/labgrid-project/labgrid/blob/f63dfd70a5da2c7007ee8c320fa969df2c0bf16e/tests/test_sigrok.py#L11-L12

JSydll commented:
Hmm, I'm actually not sure where this dependency is now coming from. It's not like the tests I want add depend on it. Maybe I forgot to inject the qemu_mock in one of the test cases? Could that be the root cause? I'll try this as soon as time permits.

JSydll committed:
test_qemudriver: Account for dependency to qemu_version_mock and avoid pulling in actual qemu-system-arm.

Signed-off-by: Joschka Seydell <joschka@seydell.org>

Bastian-Krause commented on labgrid/driver/qemudriver.py:
```diff
@@ -286,12 +288,19 @@ def on(self):
         # Restore port forwards
         for v in self._forwarded_ports.values():
-            self._add_port_forward(*v)
+            self._add_port_forward(*v) 
```
```suggestion
            self._add_port_forward(*v)
```

Bastian-Krause commented:
After looking at the QEMUDriver again, I'm not too happy about starting the qemu process in `on()` (or with this PR in `prepare()` and stopping it in `off()`. This is not the fault of this PR, though. This PR only makes it even more visible.

I'll create an alternative PR starting qemu during `on_activate()` and stopping it in `on_deactivate()`. `on()` and `off()` will interact with it via QMP only. This should cover your use case because you can call `monitor_command()` right after activation. If we agree that this is a good solution, you can rebase your other changes on top.

Bastian-Krause commented:
> After looking at the QEMUDriver again, I'm not too happy about starting the qemu process in `on()` (or with this PR in `prepare()` and stopping it in `off()`. This is not the fault of this PR, though. This PR only makes it even more visible.
> 
> I'll create an alternative PR starting qemu during `on_activate()` and stopping it in `on_deactivate()`. `on()` and `off()` will interact with it via QMP only. This should cover your use case because you can call `monitor_command()` right after activation. If we agree that this is a good solution, you can rebase your other changes on top.

See #1771.

JSydll commented:
Makes sense. I haven't expected a proposal so early, thanks!

Bastian-Krause commented:
#1771 was merged, please rebase.
